pallets/interest-accrual/src/lib.rs

```diff
@@ -132,82 +160,70 @@ pub mod pallet {
-				interest_rate_per_sec,
-				|rate_details| -> Result<T::Balance, DispatchError> {
-					let rate = if let Some(rate) = rate_details {
-						let new_accumulated_rate = Self::calculate_accumulated_rate(
-							interest_rate_per_sec,
-							rate.accumulated_rate,
-							rate.last_updated,
-						)
-						.map_err(|_| Error::<T>::DebtCalculationFailed)?;
-
-						rate.accumulated_rate = new_accumulated_rate;
-						rate.last_updated = Self::now();
-
-						rate
-					} else {
-						*rate_details = Some(RateDetails {
-							accumulated_rate: One::one(),
-							last_updated: Self::now(),
-						});
-						rate_details.as_mut().expect("RateDetails now Some. qed.")
-					};
-
-					let debt = Self::calculate_debt(normalized_debt, rate.accumulated_rate)
-						.ok_or(Error::<T>::DebtCalculationFailed)?;
-					Ok(debt)
-				},
-			)
+			if let Some(rate) = Rate::<T>::get(interest_rate_per_sec) {
+				Self::calculate_debt(normalized_debt, rate.accumulated_rate)
+					.ok_or(Error::<T>::DebtCalculationFailed.into())
+			} else {
+				Err(Error::<T>::NoSuchRate.into())
+			}
 		}
 
 		pub fn get_previous_debt(
 			interest_rate_per_sec: T::InterestRate,
 			normalized_debt: T::Balance,
 			when: Moment,
 		) -> Result<T::Balance, DispatchError> {
-			Rate::<T>::try_mutate(
-				interest_rate_per_sec,
-				|rate_details| -> Result<T::Balance, DispatchError> {
-					let rate = if let Some(rate) = rate_details {
-						let new_accumulated_rate = Self::calculate_accumulated_rate(
-							interest_rate_per_sec,
-							rate.accumulated_rate,
-							rate.last_updated,
-						)
-						.map_err(|_| Error::<T>::DebtCalculationFailed)?;
-
-						rate.accumulated_rate = new_accumulated_rate;
-						rate.last_updated = Self::now();
-
-						rate
-					} else {
-						*rate_details = Some(RateDetails {
-							accumulated_rate: One::one(),
-							last_updated: Self::now(),
-						});
-						rate_details.as_mut().expect("RateDetails now Some. qed.")
-					};
-
-					let age = Self::now().checked_sub(when).unwrap();
-					let discount = checked_pow(interest_rate_per_sec, age as usize).unwrap();
-					let previous_rate = rate.accumulated_rate.checked_div(&discount).unwrap();
-
-					let debt = Self::calculate_debt(normalized_debt, previous_rate)
-						.ok_or(Error::<T>::DebtCalculationFailed)?;
-					Ok(debt)
-				},
-			)
+			if let Some(rate) = Rate::<T>::get(interest_rate_per_sec) {
+				let now = LastUpdated::<T>::get();
+				if when > now {
+					return Err(Error::<T>::NotInPast.into());
+				}
+				let delta = now - when;
+				let rate_adjustment = checked_pow(interest_rate_per_sec, delta as usize)
+					.ok_or(ArithmeticError::Overflow)?;
+				let past_rate = rate
+					.accumulated_rate
+					.checked_div(&rate_adjustment)
+					.ok_or(ArithmeticError::Underflow)?;
+				Self::calculate_debt(normalized_debt, past_rate)
+					.ok_or(Error::<T>::DebtCalculationFailed.into())
+			} else {
+				Err(Error::<T>::NoSuchRate.into())
+			}
```

Comment:
As I mentioned in the top-level comment. I think we should note it explicitly that `past_debt` will not be correct 
`if when < last_adjusted_normalized_debt`